crates/server/src/server/h3_handler.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2015-2021 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{io, net::SocketAddr, sync::Arc};
+
+use bytes::{Buf, Bytes};
+use drain::Watch;
+use futures_util::lock::Mutex;
+use h3::server::RequestStream;
+use h3_quinn::BidiStream;
+use tracing::{debug, warn};
+use trust_dns_proto::{
+    error::ProtoError, h3::h3_server::H3Connection, http::Version, https::HttpsError, rr::Record,
+};
+
+use crate::{
+    authority::MessageResponse,
+    server::{
+        request_handler::RequestHandler, response_handler::ResponseHandler, server_future,
+        Protocol, ResponseInfo,
+    },
+};
+
+pub(crate) async fn h3_handler<T>(
+    handler: Arc<T>,
+    mut connection: H3Connection,
+    src_addr: SocketAddr,
+    _dns_hostname: Option<Arc<str>>,
+    shutdown: Watch,
+) -> Result<(), ProtoError>
+where
+    T: RequestHandler,
+{
+    // TODO: we should make this configurable
+    let mut max_requests = 100u32;
+
+    // Accept all inbound requests sent over the connection.
+    loop {
+        let (_, mut stream) = tokio::select! {
+            result = connection.accept() => match result {
+                Some(Ok(next_request)) => next_request,
+                Some(Err(err)) => {
+                    warn!("error accepting request {}: {}", src_addr, err);
+                    return Err(err);
+                }
+                None => {
+                    break;
+                }
+            },
+            _ = shutdown.clone().signaled() => {
+                // A graceful shutdown was initiated.
+                break;
+            },
+        };
+
+        let request = match stream
+            .recv_data()
+            .await
+            .map_err(|e| ProtoError::from(format!("h3 stream receive data failed: {e}")))?
+        {
+            Some(mut request) => request.copy_to_bytes(request.remaining()),
+            None => {
+                warn!("h3 stream received no data");
```

Comment:
So I'm not sure what we are supposed to do when we receive no data on an accepted stream. This probably means the stream was closed before sending any data. Shouldn't I return an error instead of just a warning here?

Reply:
I thought I left a response here. I think this is ok, it would be assumed that the next ask for data would fail if we're in a bad state. I don't know enough right now to say if this represents a bad state on the request and should therefor be closed, or if it's just a single bad request (from say a mobile device) and the next could succeed? Is that theoretically possible with H3 and Quic?

Reply:
> I think this is ok, it would be assumed that the next ask for data would fail if we're in a bad state.

We currently just continue as if nothing happened, besides the warnings.

> Is that theoretically possible with H3 and Quic?

I know in QUIC this isn't possible because opening a stream sends no data to the receiver, only sending data does (am definitely no expert in QUIC).
I know very little about the HTTP/3 spec, but as far as I digged through the code, `None` can be returned if:
- We received a `HEADER` frame, in which case we should just continue and not put out a warning. (I also noticed that we don't verify the headers in DoH3 and DoH)
- The frame was already previously received.

So actually I think we should remove that warning, which I've done now.